### PR TITLE
WIP 8.0 [REF] backup modules

### PR DESCRIPTION
**Pull request by iledarn**

### CODE REVIEW

**Comment by iledarn** [comment]
Work in progress

**Comment by iledarn** [comment]
@tarzan0820, we decide to do refactoring on your backup (saas_sysadmin_*) modules. 
- All saas_sysadmin_backup_agent are taken to the saas_server module.
- from Saas_sysadmin_backup we take core part and place it in saas_portal.
- saas_sysadmin_backup is renamed to saas_portal_backup. There is only user interface for backup in it now.
- saas_sysadmin_backup_agent_s3 is renamed to saas_portal_backup_s3.

**Comment by saltonmassally** [comment]
@iledarn thanks for informing me; I am currently doing some improvements to handle deletion of old backups so this affects that.

I have a few questions:
- What was the reasoning behind the refactoring 
-  Why are you restricting it only to the case of backing up from UI only - I think the option should exist if the user want, to have backups being taken before any modifications and deletion of client db - If we have different view on this can you move that functionality into another addon?

File: saas_server/models/saas_server.py [hunk 1]
```diff
@@ -327,3 +330,28 @@ def delete_database(self):
     def rename_database(self, new_dbname):
         openerp.service.db.exp_rename(self.name, new_dbname)
         self.name = new_dbname
+
+    @api.model
+    def _transport_backup(self, db_dump, filename=None):
```
Comment (saltonmassally) [inline saas_server/models/saas_server.py:335]: If we are moving this into the core saas_server it makes sense to maybe have a default implementation... maybe something that saves the backup to disk

**Comment by saltonmassally** [comment]
@iledarn sorry I am just seeing saas_portal_backup... please disregard point 2 above

**Comment by iledarn** [comment]
@tarzan0820 > What was the reasoning behind the refactoring
We just think that such an important function as backup should be represented in its base form in the core modules. It seems to me that you are right about implementing _transport_backup somehow

**Comment by yelizariev** [comment]
@tarzan0820 , here are my thoughts about this refactoring I asked @iledarn to do

I don't think that there should be default backup implementation. Because saving backups on the same disk would lead to disk shortage error. If we don't have default backup implementation, we cannot add interface and settings for backup to default modules (saas_portal and saas_server), because we don't want to have something, that doesn't work without extra configuration. So, we have to put "interface for backup" in a separate module and specify instruction there how to make it work. 

At the same time, moving some backup code directly to saas_portal and saas_server will not lead to any problem, while it gives some (small) advantages: 
- we don't create new two modules with technical backup features only.
- someone can create another custom module with another interface and strategy for backup. So, he will not need to copy paste some basic functions, which we put to base modules

File: saas_portal/views/saas_portal.xml [hunk 2]
```diff
@@ -35,7 +35,7 @@
             <field name="name">saas_portal.client.form</field>
             <field name="model">saas_portal.client</field>
             <field name="arch" type="xml">
-                <form string="Clients" create="false" delete="false" edit="false">
+                <form string="Clients" create="false" delete="false" edit="true">
```
Comment (saltonmassally) [inline saas_portal/views/saas_portal.xml:38]: Why are we making saas_portal.client editable again? I liked the idea of making it non-editable

File: saas_portal/models/saas_portal.py [hunk 3]
```diff
@@ -397,6 +397,29 @@ class SaasPortalDatabase(models.Model):
...
+        self.ensure_one()
+
+        state = {
+            'd': self.name,
+            'client_id': self.client_id,
+        }
+
+        url = self.server_id._request_server(path='/saas_server/backup_database', state=state, client_id=self.client_id)[0]
+        res = requests.get(url, verify=(self.server_id.request_scheme == 'https' and self.server_id.verify_ssl))
+        _logger.info('delete database: %s', res.text)
```
Comment (saltonmassally) [inline saas_portal/models/saas_portal.py]: I honestly can't remember why this logger is here but it doesn't seem appropriate maybe I meant something like

`_logger.info('backup database: %s', res.text)`

If that is the case maybe it should come before the request is actually made and not after

File: saas_portal_backup/models/saas_portal.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+class SaasPortalDatabase(models.Model):
+    _inherit = 'saas_portal.database'
+
+    backup = fields.Boolean('Backup on Modify', help="Backs up first before deleting \
+                             or upgrading", default=True)
+
+    @api.multi
+    def action_backup(self):
+        self.ensure_one()
```
Comment (saltonmassally) [inline saas_portal_backup/models/saas_portal.py]: Depending on the outcome of #171 maybe we should allow recordset and iterate over self

File: saas_portal_backup/models/saas_portal.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        self._backup()
+
+    @api.multi
+    def delete_database(self):
+        if self[0].backup:
```
Comment (saltonmassally) [inline saas_portal_backup/models/saas_portal.py]: Depending on the outcome of #171 maybe we should allow recordset and iterate over self

File: saas_portal_backup/models/saas_portal.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+            self._backup()
+        return super(SaasPortalDatabase, self).delete_database()
+
+    @api.multi
+    def upgrade_database(self):
+        if self[0].backup:
```
Comment (saltonmassally) [inline saas_portal_backup/models/saas_portal.py]: Depending on the outcome of #171 maybe we should allow recordset and iterate over self

File: saas_portal_backup/models/saas_portal.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+            self._backup()
+        return super(SaasPortalDatabase, self).upgrade_database()
+
+
+class SaasPortalClient(models.Model):
+    _inherit = 'saas_portal.client'
```
Comment (saltonmassally) [inline saas_portal_backup/models/saas_portal.py:6]: saas_portal.client inherits from saas_portal.database so shouldn't that mean that we don't have to redefine methods here?

**Comment by coodec** [comment]
Hi, sorry to comment as I don't really know (for now) the complete history and funcitionality of this.

I use this module on my implementations:
https://apps.openerp.com/apps/modules/9.0/auto_backup/

It works really great and let's me send the backup to another server.

Don't you think that maybe we just need to do a way to configure this module from saas...

Daniel.

File: saas_portal_backup/models/saas_portal.py (hunk 7 above)
Comment (iledarn) [inline saas_portal_backup/models/saas_portal.py:6]: @tarzan0820, I had tried to do it like you said. But it hadn't  worked. Then I see this post https://twitter.com/AndreasStauder/status/699506174480228352 and issue there

**Comment by yelizariev** [comment]
@iledarn consider to use Daniel's module as a base for saas_server_backup_ftp module (with keeping authorship and link to original module)

File: saas_portal_backup/models/saas_portal.py (hunk 7 above)
Comment (yelizariev) [inline saas_portal_backup/models/saas_portal.py:6]: i.e. inheriting in saas_portal.database only works only in odoo 10.0

File: saas_portal/views/saas_portal.xml (hunk 2 above)
Comment (iledarn) [inline saas_portal/views/saas_portal.xml:38]: @tarzan0820, we have new settings added on this form. Some of them (Block clients on expiration, Block clients on storage exceed) must be editable. I think also that the 'Backup on Modify' field should be editable too.

File: saas_portal_backup/models/saas_portal.py (hunk 6 above)
Comment (yelizariev) [inline saas_portal_backup/models/saas_portal.py]: overwritting of `upgrade_database` has no sense here, as this fuction only call wizard

File: saas_portal/views/saas_portal.xml (hunk 2 above)
Comment (iledarn) [inline saas_portal/views/saas_portal.xml:38]: Besides, why not do it editable if we can restrict edition for each field separately

**Pull request merged**